testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveSparkCompatibility.java

```diff
@@ -423,6 +424,78 @@ private void testReadTrinoCreatedTable(String tableName, String tableFormat)
+        onSpark().executeQuery(format("INSERT INTO %s PARTITION(dt='123-04-05') VALUES (5)", sparkTableName));
+        onSpark().executeQuery(format("INSERT INTO %s PARTITION(dt='-0001-01-01') VALUES (6)", sparkTableName));
+        onSpark().executeQuery(format("INSERT INTO %s PARTITION(dt='1965-09-10 invalid') VALUES (7)", sparkTableName));
+        onSpark().executeQuery(format("INSERT INTO %s PARTITION(dt='invalid date') VALUES (8)", sparkTableName));
+
+        assertThat(onSpark().executeQuery("SHOW PARTITIONS " + sparkTableName))
+                .containsOnly(List.of(
+                        row("dt=1965-09-10 23%3A59%3A59.999999999"),
+                        row("dt=1965-09-10 23%3A59%3A59"),
+                        row("dt=1965-09-10 23%3A59"),
+                        row("dt=12345-06-07"),
+                        row("dt=123-04-05"),
+                        row("dt=-0001-01-01"),
+                        row("dt=1965-09-10 invalid"),
+                        row("dt=invalid date")));
+
+        // Spark returns incorrect value in case of small digits
+        // Use date_format function to avoid exception due to java.sql.Date.valueOf
+        assertThat(onSpark().executeQuery("SELECT value, dt FROM " + sparkTableName))
+                .containsOnly(List.of(
+                        row(1, "1965-09-10"),
+                        row(2, "1965-09-10"),
+                        row(3, "1965-09-10"),
+                        row(4, "+12345-06-07"),
+                        row(5, null),
+                        row(6, "-0001-01-01"),
+                        row(7, "1965-09-10"),
+                        row(8, null)));
+
+        // Hive returns incorrect value for small dates
+        // Use date_format function to avoid exception due to java.sql.Date.valueOf
+        assertThat(onHive().executeQuery("SELECT value, date_format(dt, 'yyyy-MM-dd') FROM " + sparkTableName))
+                .containsOnly(List.of(
+                        row(1, "1965-09-10"),
+                        row(2, "1965-09-10"),
+                        row(3, "1965-09-10"),
+                        row(4, "12345-06-07"),
+                        row(5, "0123-04-06"),
+                        row(6, "0002-01-03"),
+                        row(7, "1965-09-10"),
+                        row(8, null)));
+
+        // Trino allows only valid format
+        // Cast to varchar so that we can compare with Spark & Hive easily
+        assertThat(onTrino().executeQuery("SELECT value, CAST(dt AS VARCHAR) FROM " + trinoTableName))
+                .containsOnly(List.of(
+                        row(1, "1965-09-10"),
+                        row(2, "1965-09-10"),
+                        row(3, "1965-09-10"),
+                        row(4, "12345-06-07"),
+                        row(5, "0123-04-05"),
+                        row(6, "-0001-01-01"),
+                        row(7, null),
+                        row(8, null)));
```

Comment:
Some values are different from Hive result (5, 6 & 7), but I hesitated to follow the implementation because it look wrong values. Let me know if we should respect Hive strictly.